### PR TITLE
Adding all keys to testdata

### PR DESCRIPTION
**Pull request by robertbastian**
`icu_testdata` currently doesn't include a handful of properties keys. It also doesn't include some collations for locales that it includes other data in.

### CODE REVIEW

**Comment by Manishearth** [comment]
Deferring to Shane on this, he has more context

File: .github/workflows/build-test.yml [hunk 1]
```diff
@@ -164,7 +164,12 @@ jobs:
     - name: Show the selected Rust toolchain
       run: rustup show
 
-    # Job specific dependencies
+    # Job-specific dependencies
+    - name: Attempt to load download cache
+      uses: actions/cache@v3
+      with:
+        key: download-cache
+        path: /tmp/icu4x
```
Comment (sffc) [inline .github/workflows/build-test.yml:172]: Nit: This is an unrelated change that should be in its own PR

File: tools/testdata-scripts/src/bin/make-testdata.rs [hunk 2]
```diff
@@ -68,13 +48,7 @@ fn main() {
...
     icu_datagen::datagen(
         Some(&icu_testdata::locales()),
-        &icu_datagen::all_keys_with_experimental()
-            .into_iter()
-            .filter(|k| !IGNORED_KEYS.contains(&&*k.path()))
-            .chain(core::iter::once(
-                icu_provider::hello_world::HelloWorldV1Marker::KEY,
-            ))
-            .collect::<Vec<_>>(),
+        &icu_datagen::all_keys_with_experimental(),
```
Comment (sffc) [inline tools/testdata-scripts/src/bin/make-testdata.rs:51]: Nit: I'd prefer to keep the hello_world key and revert the changes related to that. It has been a useful inclusion, such as the ability to use it in docstests. It's not clear that it's an improvement that we need to use some other component like icu_list or pull data from the HelloWorldProvider instead of the testdata provider.

File: provider/blob/benches/provider_blob.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use criterion::{black_box, criterion_group, criterion_main, Criterion};
+
+use icu_locid::langid;
+use icu_provider::hello_world::HelloWorldV1Marker;
+use icu_provider::prelude::*;
+use icu_provider_blob::BlobDataProvider;
+
+static BLOB: &[u8] = include_bytes!(concat!(
```
Comment (sffc) [inline provider/blob/benches/provider_blob.rs:12]: Nit: The changes to benches should really be in their own PR. A PR entitled "adding all keys to testdata" should not have earmarks in it.

**Review by sffc — CHANGES_REQUESTED** [review]

File: .github/workflows/build-test.yml (hunk 1 above)
Comment (robertbastian) [reply .github/workflows/build-test.yml:172]: It's not. This PR builds testdata from a tag instead of the locally downloaded subset of sources.

File: tools/testdata-scripts/src/bin/make-testdata.rs (hunk 2 above)
Comment (robertbastian) [reply tools/testdata-scripts/src/bin/make-testdata.rs:51]: I'll split it out to discuss further

File: provider/blob/benches/provider_blob.rs (hunk 3 above)
Comment (robertbastian) [reply provider/blob/benches/provider_blob.rs:12]: This is a consequence of the hello world removal

File: tools/testdata-scripts/src/bin/make-testdata.rs [hunk 4]
```diff
@@ -34,9 +14,9 @@ fn main() {
         .unwrap();
 
     let source_data = SourceData::default()
-        .with_cldr(paths::cldr_json_root(), CldrLocaleSubset::Full)
+        .with_cldr_for_tag(&icu_testdata::versions::cldr_tag(), CldrLocaleSubset::Full)
```
Comment (sffc) [inline tools/testdata-scripts/src/bin/make-testdata.rs:17]: Issue: Please, please make this read from the repo. A dev should be able to run `cargo make testdata` without an internet connection. This has always worked and there's no reason to break it.

**Review by sffc — CHANGES_REQUESTED** [review]

File: tools/testdata-scripts/src/bin/make-testdata.rs (hunk 4 above)
Comment (robertbastian) [reply tools/testdata-scripts/src/bin/make-testdata.rs:17]: The repo might not reflect the correct data for this tag. This is more of an issue for the follow-up PR, where we generate testdata at `SourceData::LATEST_TESTED_CLDR_TAG` and then write that tag into the metadata. Local data could be wrong because someone forgot a glob, or we just haven't run the download script with the latest tag. I'm also more confident calling the constant `LATEST_TESTED_CLDR_TAG` if we actually test with it.

"Download" has been required for many things locally and on CI for quite a while now. Because it caches it's a one-time thing and we haven't had issues with it.

Comment (sffc) [reply tools/testdata-scripts/src/bin/make-testdata.rs:17]: The solution we've discussed is that there should be a CI action that verifies that the repo data is not stale.

> "Download" has been required for many things locally and on CI for quite a while now.

CI yes, but we've kept local development as no-network as possible.

It's also much easier to reason about the data pipeline if I can see exactly what data is being consumed by datagen in order to produce the processed ICU4X output data. Depending on some artifact downloaded from the network is simply not a good idea for the purposes of ICU4X development.

File: tools/testdata-scripts/src/bin/make-testdata.rs [hunk 5]
```diff
@@ -34,9 +14,9 @@ fn main() {
...
         .unwrap()
-        .with_icuexport(paths::icuexport_toml_root())
+        .with_icuexport_for_tag(&icu_testdata::versions::icu_tag())
```
Comment (sffc) [inline tools/testdata-scripts/src/bin/make-testdata.rs:19]: The title of this PR should be "Make testdata datagen download its sources directly"; that's the bigger change being made. The fact that it gets a few more keys is a side-effect.

**Review by sffc — CHANGES_REQUESTED** [review]

File: tools/testdata-scripts/src/bin/make-testdata.rs (hunk 4 above)
Comment (robertbastian) [reply tools/testdata-scripts/src/bin/make-testdata.rs:17]: I'd argue relying on the network is much less of a problem locally than in CI.  Then work-on-plane use case is rare, and even then you're gonna have the data in the cache unless you started ICU4X development just before boarding.

Comment (robertbastian) [reply tools/testdata-scripts/src/bin/make-testdata.rs:17]: Do you disagree with my point about calling the latest-tested-tag tested?

**Comment by sffc** [comment]
Also remember that I was reluctant to go along with the user-facing version of icu4x-datagen downloading its data automatically; I agreed to that after weighing the following advantages: (1) icu4x-datagen has a lot of flags and we want to make it more accessible, since it's one of the first things ICU4X users need to use, and (2) as we add more data sources, we need to keep older invocations of icu4x-datagen working properly, which means we need to give them default sources.

Neither of those points for icu4x-datagen applies here in testdata datagen.

**Comment by robertbastian** [comment]
What's your preferred solution to the final testdata script then? 

https://github.com/unicode-org/icu4x/blob/c2a5392ca8cdba3b4a89eb177b820fe911f1342e/tools/testdata-scripts/src/bin/make-testdata.rs#L27-L90

Note that we generate testdata for a tag and then write that tag to metadata. Do we just pray that the local data is what would have been downloaded for the tag?

**Comment by sffc** [comment]
There should be a CI that verifies that the copy of the data in the repo corresponds to the tag.

**Comment by sffc** [comment]
See https://github.com/unicode-org/icu4x/issues/2880

**Comment by robertbastian** [comment]
This is where I fundamentally disagree, keeping state in the repo and having CI verify that the state is correct adds a lot of friction when the state is easy to regenerate. I already filed #3006 about this.

**Comment by robertbastian** [comment]
Also the data will not correspond to the tag, because we only include select files. That itself could hide bugs.

**Comment by sffc** [comment]
> This is where I fundamentally disagree, keeping state in the repo and having CI verify that the state is correct adds a lot of friction when the state is easy to regenerate. I already filed #3006 about this.

I see the friction argument for things that change "unexpectedly", like the GN files. But, these are files that don't change often, and when they do, it's normally _intentional_. If they get out of sync, that's a _mistake_, not an _oversight_.

> Also the data will not correspond to the tag, because we only include select files. That itself could hide bugs.

Can you elaborate on the types of bugs this creates?

**Comment by robertbastian** [comment]
> If they get out of sync, that's a mistake, not an oversight.

I don't understand the difference and how that changes things.

> Can you elaborate on the types of bugs this creates?

A new file gets added to icuexport but not to the glob. This file breaks some assumption in datagen (which often lists directories) and breaks it.

**Review by sffc — APPROVED** [review]
It doesn't make testdata smaller, but if adding all keys is the desired outcome, I like this way of doing it

**Comment by robertbastian** [comment]
Testdata should contain all keys so that it works with all constructors. For collation maybe it's including too much, but we have to eat our own dogfood here, datagen users cannot at the moment exclude things like `und-u-co-emoji`.
